File: cpp/problem024.cpp
#include <iostream>
#include <vector>


using namespace std;


int factorial(int n)
{
    int f = 1, i = 1;
    while (++i <= n) f *= i;
    return f;
}


vector<int> factorialization(int n) {
    int k = 0;

    while (factorial(++k) < n);

    vector<int> v;
    while (k-- > 0) {
        int a = n / factorial(k);
        n -= a * factorial(k);
        v.push_back(a);
    }

    return v;
}


int main()
{
    int s = 10, n = 1000000;

    vector<int> S;
    for (int i = 0; i < s; ++i) S.push_back(i);

    for (int x: factorialization(n - 1)) {
        cout << S[x] << " ";
        S.erase(S.begin() + x);
    }

    cout << endl;
    return 0;
}

File: python/problem024.py
def factorial(n):
     ''' Compute n!
     '''
     if n == 0: return 1
     return factorial(n-1) * n


def factorialization(n):
     ''' f(n) = X : (sum(x!) for each x in X) == n.
     '''
     k = 1
     while factorial(k) < n:
         k += 1

     factors = []
     while k > 0:
         k -= 1
         a = n // factorial(k)
         b = n - a * factorial(k)
         factors.append(a)
         n = b

     return factors


def nth_permutation(s, n):
     ''' return n-th lexicographic permutation of s elements <0..s)
     '''
     S = [i for i in range(s)]
     p = []

     for x in factorialization(n - 1):
         p.append(S[x])
         del S[x]

     return p


print(nth_permutation(10, 1000000))
